src/deploy/functions/cache/applyHash.ts

```diff
@@ -1,30 +1,33 @@
 import { Backend, allEndpoints } from "../backend";
 import * as args from "../args";
-import {
-  getEndpointHash,
-  getEnvironmentVariablesHash,
-  getSecretsHash,
-  getSourceHash,
-} from "./hash";
+import { getEndpointHash, getEnvironmentVariablesHash, getSecretsHash } from "./hash";
+import { EndpointFilter, isCodebaseFiltered, isEndpointFiltered } from "../functionsDeployHelper";
 
 /**
  *
  * Updates all the CodeBase {@link Backend}, applying a hash to each of their {@link Endpoint}.
  */
-export async function applyBackendHashToBackends(
+export function applyBackendHashToBackends(
   wantBackends: Record<string, Backend>,
   context: args.Context
-): Promise<void> {
+): void {
   for (const [codebase, wantBackend] of Object.entries(wantBackends)) {
+    // If an entire codebase is filtered, then don't set the hash for the functions in the codebase.
+    // This effectively forces all the functions to deploy without the duplication check.
+    if (isCodebaseFiltered(codebase, context.filters || [])) {
+      continue;
+    }
     const source = context?.sources?.[codebase]; // populated earlier in prepare flow
-    const sourceV1Hash = source?.functionsSourceV1
-      ? await getSourceHash(source?.functionsSourceV1)
-      : undefined;
-    const sourceV2Hash = source?.functionsSourceV2
-      ? await getSourceHash(source?.functionsSourceV2)
-      : undefined;
     const envHash = getEnvironmentVariablesHash(wantBackend);
-    applyBackendHashToEndpoints(wantBackend, envHash, sourceV1Hash, sourceV2Hash);
+    const filtersFilteredByCodebase =
+      context.filters?.filter((filter) => filter.codebase === codebase) || [];
```

Comment:
IIRC, when codebase === undefined, it should match the idchunks against all codebases:

https://github.com/firebase/firebase-tools/blob/2f8f0cb2f1660c18ca888c37ef705cc558147114/src/deploy/functions/functionsDeployHelper.ts#L5-L7

We should either include filters whose codebase is undefined or... not do this filtering at all? I'm not 100% sure if this filtering is necessary.

Reply:
Thats a very good point (and it also causes the bug that you mentioned.) Removed the filtered filter, which fixed the bug (verified with TDD).